mllib/src/main/scala/org/apache/spark/mllib/evaluation/MulticlassMetrics.scala

```diff
@@ -39,21 +46,28 @@ class MulticlassMetrics @Since("1.1.0") (predictionAndLabels: RDD[(Double, Doubl
   private[mllib] def this(predictionAndLabels: DataFrame) =
     this(predictionAndLabels.rdd.map(r => (r.getDouble(0), r.getDouble(1))))
 
-  private lazy val labelCountByClass: Map[Double, Long] = predictionAndLabels.values.countByValue()
-  private lazy val labelCount: Long = labelCountByClass.values.sum
-  private lazy val tpByClass: Map[Double, Int] = predictionAndLabels
-    .map { case (prediction, label) =>
-      (label, if (label == prediction) 1 else 0)
+  private lazy val labelCountByClass: Map[Double, Double] =
+    predLabelsWeight.map {
+      case (prediction: Double, label: Double, weight: Double) =>
+        (label, weight)
+    }.reduceByKey(_ + _).collect().toMap
+  private lazy val labelCount: Double = labelCountByClass.values.sum
+  private lazy val tpByClass: Map[Double, Double] = predLabelsWeight
+    .map {
```

Comment:
I'm not sure, but I wonder if it's more efficient to filter on label == prediction, then reduce? The original code didn't do it, but could be worth it.

Reply:
done, not sure if more efficient, it seemed to take same time for me, but I didn't test extensively

Reply:
it looks like this will actually cause tests to fail, because the key may become missing if we filter everything out first, whereas we would want it to be present otherwise but have a 0 value

Reply:
see the test failure here for reference: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/98529/testReport/